protocols/kad/src/query/peers/closest/disjoint.rs

```diff
@@ -0,0 +1,965 @@
+    /// containing the corresponding iterator indices as well as the response
+    /// state.
+    ///
+    /// Used to track which iterator contacted which peer. See [`PeerState`]
+    /// for details.
+    contacted_peers: HashMap<PeerId, PeerState>,
+}
+
+impl ClosestDisjointPeersIter {
+    /// Creates a new iterator with a default configuration.
+    pub fn new<I>(target: KeyBytes, known_closest_peers: I) -> Self
+    where
+        I: IntoIterator<Item = Key<PeerId>>,
+    {
+        Self::with_config(
+            ClosestPeersIterConfig::default(),
+            target,
+            known_closest_peers,
+        )
+    }
+
+    /// Creates a new iterator with the given configuration.
+    pub fn with_config<I, T>(
+        config: ClosestPeersIterConfig,
+        target: T,
+        known_closest_peers: I,
+    ) -> Self
+    where
+        I: IntoIterator<Item = Key<PeerId>>,
+        T: Into<KeyBytes> + Clone,
+    {
+        let peers = known_closest_peers.into_iter().take(K_VALUE.get()).collect::<Vec<_>>();
+        let iters = (0..config.parallelism.get())
+            // NOTE: All [`ClosestPeersIter`] share the same set of peers at
+            // initialization. The [`ClosestDisjointPeersIter.contacted_peers`]
+            // mapping ensures that a successful response from a peer is only
+            // ever passed to a single [`ClosestPeersIter`]. See
+            // [`ClosestDisjointPeersIter::on_success`] for details.
+            .map(|_| ClosestPeersIter::with_config(config.clone(), target.clone(), peers.clone()))
+            .collect::<Vec<_>>();
+
+        let iters_len = iters.len();
+
+        ClosestDisjointPeersIter {
+            config,
+            target: target.into(),
+            iters,
+            iter_order: (0..iters_len).map(IteratorIndex as fn(usize) -> IteratorIndex).cycle(),
+            contacted_peers: HashMap::new(),
+        }
+    }
+
+    /// Callback for informing the iterator about a failed request to a peer.
+    ///
+    /// If the iterator is currently waiting for a result from `peer`,
+    /// the iterator state is updated and `true` is returned. In that
+    /// case, after calling this function, `next` should eventually be
+    /// called again to obtain the new state of the iterator.
+    ///
+    /// If the iterator is finished, it is not currently waiting for a
+    /// result from `peer`, or a result for `peer` has already been reported,
+    /// calling this function has no effect and `false` is returned.
+    pub fn on_failure(&mut self, peer: &PeerId) -> bool {
+        let mut updated = false;
+
+        // All peer failures are reported to all queries and thus to all peer
+        // iterators. If this iterator never started a request to the given peer
+        // ignore the failure.
+        if let Some(PeerState{ initiated_by, response }) = self.contacted_peers.get_mut(peer) {
+            *response = ResponseState::Failed;
```

Comment:
Good catch. Thanks!

> I guess, however, that it may be difficult to work around the need to ask for
  this state from another iterator while (mutably) iterating over the iterators
  in next()?

Yes, I am unable to come up with a solution for `next()` without introducing
unreasonable complexity.

<details>
  <summary>Double borrow error</summary>
  
  ```
error[E0502]: cannot borrow `self.iters` as immutable because it is also borrowed as mutable
   --> protocols/kad/src/query/peers/closest/disjoint.rs:211:39
    |
177 |             let iter = &mut self.iters[i];
    |                             ---------- mutable borrow occurs here
...
180 |                 match iter.next(now) {
    |                       ---- mutable borrow later used here
...
211 |                                 match self.iters[*initiated_by].status(peer) {
    |                                       ^^^^^^^^^^ immutable borrow occurs here
  ```
</details>

What do you think of https://github.com/libp2p/rust-libp2p/pull/1473/commits/4955330a15b746209ad221b6eb9f1e4796ea8488?

In case you don't have any objections I will deploy this to kademlia-exporter.max-inden.de. This might explain the increased number of connections per query.